Wikipedia/Code/WMFArticleViewController.m

```diff
@@ -95,13 +95,14 @@ - (nullable UIActivityViewController *)sharingActivityViewControllerWithTextSnip
     [items addObject:[[WMFArticleTextActivitySource alloc] initWithArticle:self shareText:text]];
 
     NSURL *url = [NSURL wmf_desktopURLForURL:self.url];
-
+    
     if (url) {
-        NSString *urlAsString = url.absoluteString;
-        if([urlAsString hasSuffix:@"!"]) {
-            urlAsString = [urlAsString stringByReplacingOccurrencesOfString:@"!" withString:@"%21"];
-        }
-        [items addObject:[[NSURL alloc] initWithString:[NSString stringWithFormat:@"%@?%@", urlAsString, @"wprov=sfsi1"]]];
+        NSURLComponents *components = [NSURLComponents componentsWithURL:url resolvingAgainstBaseURL:NO];
+        
+        NSURLQueryItem *queryItem = [NSURLQueryItem queryItemWithName: @"wprov" value: @"sfsi1"];
+        components.queryItems = @[queryItem];
+
+        [items addObject:components.URL];
```

Comment:
`components.URL` could be nil (the property on `NSURLComponents` is declared as `nullable`, also `components` itself could be nil) and adding nil to an array causes a crash. This just needs to be assigned to a variable and nil checked before adding it to the array. Otherwise looks good!